### ISSUE TITLE
sql: panic: fatal error: stack overflow

### ISSUE THREAD

**Issue by mjibson** [cockroachdb/cockroach#41772]
```
CREATE TABLE table1 ();

WITH
        with_5211 AS (SELECT NULL FROM table1),
        with_5212
                AS (
                        SELECT
                                *
                        FROM
                                (VALUES (NOT NULL, ARRAY[0, 0, 0, 0:::OID]))
                                        AS tab_12862 (col_29995, col_29996)
                        UNION ALL
                                SELECT
                                        *
                                FROM
                                        (
                                                VALUES
                                                        (NULL, NULL),
                                                        (false, ARRAY[0:::OID])
                                        )
                                                AS tab_12863 (col_29997, col_29998)
                )
SELECT
        NULL
FROM
        with_5211, with_5212
WHERE
        with_5212.col_29995;
```

```
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*Factory).ConstructInnerJoin(0xc000ad4a98, 0x48b59c0, 0xc001531158, 0x48b54c0, 0xc0003e3818, 0xc001843680, 0x0, 0x1, 0xc0018b2410, 0x0, ...)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:1651 +0x6dda fp=0xc024000660 sp=0xc024000658 pc=0x178ff6a
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*Factory).ConstructInnerJoin(0xc000ad4a98, 0x48b59c0, 0xc001531158, 0x48b4d40, 0xc0018b2198, 0xc001843680, 0x0, 0x1, 0xc0018b2410, 0xc0003e3818, ...)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:2370 +0x2ea4 fp=0xc024000de8 sp=0xc024000660 pc=0x178c034
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*Factory).DynamicConstruct(0xc000ad4a98, 0xc00120005c, 0xc001204680, 0x4, 0x4, 0x48b4d40, 0xc0018b2198)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:16979 +0x471d fp=0xc024000f88 sp=0xc024000de8 pc=0x180972d
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*Factory).ConstructProject(0xc000ad4a98, 0x48b43e0, 0xc0018b23d8, 0xc001016000, 0x0, 0x1, 0x0, 0x0, 0xc0018b2410, 0x48b43e0)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:1316 +0x3542 fp=0xc024001530 sp=0xc024000f88 pc=0x17870b2
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*Factory).ConstructInnerJoin(0xc000ad4a98, 0x48b59c0, 0xc001531158, 0x48b4d40, 0xc0018b2198, 0xc001843680, 0x0, 0x1, 0xc0018b2410, 0xc0003e3818, ...)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:2369 +0x2f69 fp=0xc024001cb8 sp=0xc024001530 pc=0x178c0f9
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*Factory).DynamicConstruct(0xc000ad4a98, 0xc00120005c, 0xc001204640, 0x4, 0x4, 0x48b4d40, 0xc0018b2198)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:16979 +0x471d fp=0xc024001e58 sp=0xc024001cb8 pc=0x180972d
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*Factory).ConstructProject(0xc000ad4a98, 0x48b43e0, 0xc0018b23d8, 0xc001016000, 0x0, 0x1, 0x0, 0x0, 0xc0018b2410, 0x48b43e0)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:1316 +0x3542 fp=0xc024002400 sp=0xc024001e58 pc=0x17870b2
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*Factory).ConstructInnerJoin(0xc000ad4a98, 0x48b59c0, 0xc001531158, 0x48b4d40, 0xc0018b2198, 0xc001843680, 0x0, 0x1, 0xc0018b2410, 0xc0003e3818, ...)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:2369 +0x2f69 fp=0xc024002b88 sp=0xc024002400 pc=0x178c0f9
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*Factory).DynamicConstruct(0xc000ad4a98, 0xc00120005c, 0xc001204600, 0x4, 0x4, 0x48b4d40, 0xc0018b2198)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:16979 +0x471d fp=0xc024002d28 sp=0xc024002b88 pc=0x180972d
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*Factory).ConstructProject(0xc000ad4a98, 0x48b43e0, 0xc0018b23d8, 0xc001016000, 0x0, 0x1, 0x0, 0x0, 0xc0018b2410, 0x48b43e0)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:1316 +0x3542 fp=0xc0240032d0 sp=0xc024002d28 pc=0x17870b2
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*Factory).ConstructInnerJoin(0xc000ad4a98, 0x48b59c0, 0xc001531158, 0x48b4d40, 0xc0018b2198, 0xc001843680, 0x0, 0x1, 0xc0018b2410, 0xc0003e3818, ...)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:2369 +0x2f69 fp=0xc024003a58 sp=0xc0240032d0 pc=0x178c0f9
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*Factory).DynamicConstruct(0xc000ad4a98, 0xc00120005c, 0xc0012045c0, 0x4, 0x4, 0x48b4d40, 0xc0018b2198)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:16979 +0x471d fp=0xc024003bf8 sp=0xc024003a58 pc=0x180972d
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*Factory).ConstructProject(0xc000ad4a98, 0x48b43e0, 0xc0018b23d8, 0xc001016000, 0x0, 0x1, 0x0, 0x0, 0xc0018b2410, 0x48b43e0)
...
```

**Comment by RaduBerinde** [comment]
It looks like the `HoistJoinProjectRight` rule, @justinj I think you did a bit of work related to that (you added the "left" version), can you take a look?

**Comment by justinj** [comment]
Yes, will do.

**Comment by justinj** [comment]
I think the issue with this is that our `DerivePruneCols` for `UnionAll` advertises prune columns that it won't follow through with sometimes (because we unconditionally advertise all output cols and only push down if both our inputs can be pruned. I have a fix for this but it triggers #41973, so I will wait on that for the time being.

**Comment by RaduBerinde** [comment]
Fixed in https://github.com/cockroachdb/cockroach/pull/41984.
